[WIP] Docstring for public methods in MatrixBase

kritkaran94 opened the pull request:
This PR addresses issue #11344. @leosartaj so far I have added docstring for inv and trace module. I will add for others too.

kritkaran94 committed:
Merge pull request #1 from kritkaran94/#11245

Issue #11245

kritkaran94 committed:
added docstring for inv, trace module

kritkaran94 committed:
removed trailing white space

kritkaran94 committed:
updated docstring

leosartaj commented on sympy/matrices/matrices.py:
```diff
@@ -302,11 +302,46 @@ def copy(self):
+        >>> from sympy import MatrixSymbol, Trace, eye
+        >>> A = MatrixSymbol('A', 3, 3)
+        >>> Trace(A)
+        Trace(A)
```
You need to give examples for ``trace`` and not ``Trace``.

leosartaj commented on sympy/matrices/matrices.py:
```diff
@@ -302,11 +302,46 @@ def copy(self):
 
     def trace(self):
+        """
+        Represents the trace of a matrix expression.
```
Returns the trace of a matrix expression.

leosartaj commented on sympy/matrices/matrices.py:
```diff
@@ -302,11 +302,46 @@ def copy(self):
     def inv(self, method=None, **kwargs):
+        """
+        Returns the inverse of the matrix `K`, if it exists using the Gaussian elimination by default.
+        However we can specify it to be done by LU decomposition as well.
```
What happens when ``inv`` does not exist. Write a line about that as well.

leosartaj replied:
Returns the inverse of a matrix , using Gaussian elimination by default. By setting ``method='LU'``  we can specify it to be done using LU decomposition.

kritkaran94 commented:
@leosartaj I will make the changes and get back to you.

kritkaran94 committed:
Fixes Issue #11344

kritkaran94 commented:
@leosartaj I have made the changes. However, travis fails and points out to the file I din't change (sparse.py). Any idea on this ?

kritkaran94 committed:
added docstring for inv, trace, copy etc..

asmeurer commented on sympy/matrices/matrices.py:
```diff
@@ -299,23 +299,66 @@ def _setitem(self, key, value):
 
     def trace(self):
+        """
+        Return the trace of a matrix expression ie.. the sum of the
```
i.e.

kritkaran94 replied:
updated ie.. --> ie. (Thanks for that)

asmeurer commented on sympy/matrices/matrices.py:
```diff
@@ -299,23 +299,66 @@ def _setitem(self, key, value):
+
+        Also, by setting method='LU' we can specify it to be done using LU decomposition.
+
+        However, if determinant of matrix is zero, a value error is returned as the inverse for that
```
"a ValueError is raised"

kritkaran94 replied:
Updated from value error --> ValueError

kritkaran94 committed:
updated docstring for trace, inv module

kritkaran94 committed:
trailing white space error fixed

kritkaran94 commented:
@leosartaj All the tests have passed and I have made a couple of changes pointed out by you and @asmeurer. If there's anything else to add just let me know.

If it looks good to go, then it can be merged?

leosartaj commented on sympy/matrices/matrices.py:
```diff
@@ -299,23 +299,66 @@ def _setitem(self, key, value):
 
     def copy(self):
+        """
+        Return the copy of a matrix expression.
```
returns

leosartaj commented on sympy/matrices/matrices.py:
```diff
@@ -299,23 +299,66 @@ def _setitem(self, key, value):
 
     def trace(self):
+        """
+        Return the trace of a matrix expression ie. the sum of the
```
returns, i.e.

leosartaj commented on sympy/matrices/matrices.py:
```diff
@@ -299,23 +299,66 @@ def _setitem(self, key, value):
     def trace(self):
+        """
+        Return the trace of a matrix expression ie. the sum of the
+        elements on the diagonal.
```
`elements on the diagonals` -> `diagonal elements`

leosartaj commented on sympy/matrices/matrices.py:
```diff
@@ -299,23 +299,66 @@ def _setitem(self, key, value):
+
+        Also, by setting method='LU' we can specify it to be done using LU decomposition.
+
+        However, if determinant of matrix is zero, a ValueError is raised as the inverse for that
```
You should wrap around code things  with \`\` for eg.  ``ValueError``

leosartaj replied:
determinant of the matrix.

kritkaran94 replied:
@leosartaj Will make the changes right away.

kritkaran94 replied:
updated ie. -> i.e.
updated ValueError --> ``ValueError``
updated Return --> Returns
updated on the diagonals -> diagonal elements

kritkaran94 committed:
fixed issues w.r.t articles

kritkaran94 commented:
@leosartaj Any other changes? I made note of all the previous comments.

kritkaran94 commented:
ping, @leosartaj

kritkaran94 commented:
@jksuom Could you please review this PR ?

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -299,22 +299,65 @@ def _setitem(self, key, value):
+
+        Also, by setting method='LU' we can specify it to be done using LU decomposition.
+
+        However, if determinant of the matrix is zero, a ``ValueError`` is raised as the inverse for that
```
if determinant -> if the determinant

This line and the one above appear too long for an 80 character terminal.

kritkaran94 replied:
@jksuom, in the previous comments I was asked to add details about what would happen if inverse doesn't exist and also write something about `method = LU`. Do you want it to be rephrased or removed ?


I apologize for replying so late.

jksuom commented:
There is another `trace` for matrix expressions that will typically produce a `Trace` object. It seems that this trace is for actual matrices. So the docstring should probably say just 'matrix' instead of 'matrix expression'. The same in `copy`. Otherwise, this looks good.

kritkaran94 committed:
updated docstring for trace, copy

kritkaran94 commented:
@jksuom The docstring for ``trace`` and ``copy`` have now been updated.

kritkaran94 commented:
@jksuom Does it look good?

jksuom commented:
The first line of the docstring of `inv` could be simply
`Return the inverse of a matrix.`

Then, after a blank line, you can explain that there are two methods, Gaussian elimination and LU, with the former being the default.

Finally, there should be the underlined heading `Raises` and the explanation of `ValueError` thereafter.

Also, pay attention to the line length. It should preferably not exceed 72 characters.

kritkaran94 committed:
rectified character length in inv()

kritkaran94 commented:
The ``inv`` module has now been modified. I have tried to cut short the character length.

````
        Return the inverse of a matrix.

        Inverse is found using Gaussian elimination & LU, with former being default.

        Raises
        ========

        If the determinant of the matrix is zero, a ``ValueError`` is raised.

        See Also
        ========

        inverse_LU
        inverse_GE
        inverse_ADJ

````

I have rectified the usage of articles with Grammarly tool.

kritkaran94 commented:
@jksuom Any further changes ?

jksuom commented:
It seems that there are several possibilities for the method, different for dense and sparse matrices. There are also other keywords. Perhaps the various possibilities should be investigated (from dense.py and sparse.py), and the results explained here in one place.

The Raises section is usually formatted a little differently. See, for example, the Raises section [here](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt#sections)

kritkaran94 committed:
reformatted Raises section

Still several changes need to be made in the inv method as mentioned by @jksuom.

However, the changes w.r.t formatting of Raises have been made.

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -340,9 +340,9 @@ def inv(self, method=None, **kwargs):
-        If the determinant of the matrix is zero, a ``ValueError`` is raised.
+        ------
+        ValueError
+        If the determinant of the matrix is zero.
```
This line is supposed to be intended (4 spaces).

kritkaran94 replied:
Fixed that. I have also added other methods after referring to ``dense.py`

````
Return the inverse of a matrix (default is Gauss elimination).

        kwargs
        ======

        method : ('GE', 'LU', or 'ADJ')

        Notes
        =====

        According to the ``method`` keyword, it calls the appropriate method:

          GE .... inverse_GE(); default
          LU .... inverse_LU()
          ADJ ... inverse_ADJ()

        Raises
        ------
        ValueError
            If the determinant of the matrix is zero.

        See Also
        ========

        inverse_LU
        inverse_GE
        inverse_ADJ

````

kritkaran94 replied:
@jksuom Does it Look Good ?

I removed some lines that were there in the ``def _eval_inverse`` ( ``dense.py``) as ``try_block_diag`` and ``iszerosfunc`` are not there in the code under ``inv`` ( ``matrices.py``).

jksuom replied:
It should somehow be indicated that the methods are different for sparse and dense matrices. Also their defaults are different. Otherwise, it looks good.

So I think that no default should be mentioned in the short summary on the first line, and the the methods should be separated into two groups according to the type of the matrix.

kritkaran94 replied:
Okay I will do that and get back to you.

kritkaran94 committed:
added details about all inv methods

Also fixed indentation issue.

kritkaran94 committed:
added cases for dense and sparse matrix in inv()

kritkaran94 commented:
@jksuom The following changes have been made in the ``inv`` method in ``matrices.py``.

Return the inverse of a matrix.

````
        CASE 1: If the matrix is a dense matrix.

        Return the matrix inverse using the method indicated (default
        is Gauss elimination).

        kwargs
        ======

        method : ('GE', 'LU', or 'ADJ')

        Notes
        =====

        According to the ``method`` keyword, it calls the appropriate method:

          GE .... inverse_GE(); default
          LU .... inverse_LU()
          ADJ ... inverse_ADJ()

        See Also
        ========

        inverse_LU
        inverse_GE
        inverse_ADJ

        Raises
        ------
        ValueError
            If the determinant of the matrix is zero.

        CASE 2: If the matrix is a sparse matrix.

        Return the matrix inverse using Cholesky or LDL (default).

        kwargs
        ======

        method : ('CH', 'LDL')

        Notes
        =====

        According to the ``method`` keyword, it calls the appropriate method:

          LDL .... inverse_LDL(); default
          CH .... inverse_CH()

        Raises
        ------
        ValueError
            If the determinant of the matrix is zero.
````
I added two cases, it looks readable in this way?

kritkaran94 commented:
@jksuom I have made several changes. Any suggestions on the same ?

jksuom commented on sympy/matrices/matrices.py:
```diff
@@ -299,22 +299,106 @@ def _setitem(self, key, value):
+
+        According to the ``method`` keyword, it calls the appropriate method:
+
+          LDL .... inverse_LDL(); default
```
With only three dots this would be aligned with the next line.

kritkaran94 replied:
You meant `` CH ... inverse_CH() `` ?

kritkaran94 replied:
@jksuom, is this okay ?

Notes for dense matrix

````
Notes
        =====

        According to the ``method`` keyword, it calls the appropriate method:

          GE .... inverse_GE(); default
          LU .... inverse_LU()
          ADJ ... inverse_ADJ()

````
Notes for sparse matrix

````
Notes
        =====

        According to the ``method`` keyword, it calls the appropriate method:

          LDL .... inverse_LDL(); default
          CH ... inverse_CH()
````

Earlier there were issues with the alignment. Now, for the last one in each case 3 dots are there.

jksuom replied:
The other way round. If there are 3 dots in the LDL line and 4 dots on the CH line, then the texts should become aligned.

kritkaran94 replied:
okay I now get it.

kritkaran94 committed:
alignment issue for sparse in inv method

kritkaran94 committed:
alignment issue for sparse in inv()

kritkaran94 commented:
@jksuom I have fixed that. All good with the docstring for ``matrices.py``?

Ready for Merge ?

jksuom commented:
Thanks. It looks better now.

kritkaran94 commented:
Thank you for your guidance.

kritkaran94 commented:
@jksuom all the tests have passed. If there are any other changes please let me know. If it lookes good to you please merge this.
